In [6]:
import pandas as pd 
purchase_df = pd.read_csv('Data Set/QVI_purchase_behaviour.csv')
trans_df = pd.read_csv('Data Set/QVI_transaction_data.csv')

FileNotFoundError: [Errno 2] No such file or directory: 'Data Set/QVI_purchase_behaviour.csv'

# STEP 1: UNDERSTANDING THE PROBLEM

#### Goal of the Problem 

It is **Descriptive and Diagnostic** Problem with segmentation recommendations

In [ ]:
purchase_df

LYLTY_CARD_NBR               LIFESTAGE PREMIUM_CUSTOMER
0                1000   YOUNG SINGLES/COUPLES          Premium
1                1002   YOUNG SINGLES/COUPLES       Mainstream
2                1003          YOUNG FAMILIES           Budget
3                1004   OLDER SINGLES/COUPLES       Mainstream
4                1005  MIDAGE SINGLES/COUPLES       Mainstream
...               ...                     ...              ...
72632         2370651  MIDAGE SINGLES/COUPLES       Mainstream
72633         2370701          YOUNG FAMILIES       Mainstream
72634         2370751          YOUNG FAMILIES          Premium
72635         2370961          OLDER FAMILIES           Budget
72636         2373711   YOUNG SINGLES/COUPLES       Mainstream

[72637 rows x 3 columns]

#### Real meaning of the variables

Dataset 1: Purchase Behaviour Data (Customer Segments)

| Column Name        | Description |
|--------------------|-------------|
| `LYLTY_CARD_NBR`   | Unique customer **loyalty card number**. Identifies the customer and is used to link transactions to behavior. |
| `LIFESTAGE`        | Describes the life stage/demographic of the customer such as "YOUNG FAMILIES", "MIDAGE SINGLES/COUPLES", etc. |
| `PREMIUM_CUSTOMER` | Indicates the customer’s spending tier:<br>• **Premium** – High spender<br>• **Mainstream** – Average spender<br>• **Budget** – Price-conscious customer |

---




In [ ]:
trans_df

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0       43390          1            1000       1         5   
1       43599          1            1307     348        66   
2       43605          1            1343     383        61   
3       43329          2            2373     974        69   
4       43330          2            2426    1038       108   
...       ...        ...             ...     ...       ...   
264831  43533        272          272319  270088        89   
264832  43325        272          272358  270154        74   
264833  43410        272          272379  270187        51   
264834  43461        272          272379  270188        42   
264835  43365        272          272380  270189        74   

                                       PROD_NAME  PROD_QTY  TOT_SALES  
0         Natural Chip        Compny SeaSalt175g         2        6.0  
1                       CCs Nacho Cheese    175g         3        6.3  
2         Smiths Crinkle Cut  Chips Chicken 170g         2        2.9  
3         Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0  
4       Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8  
...                                          ...       ...        ...  
264831   Kettle Sweet Chilli And Sour Cream 175g         2       10.8  
264832             Tostitos Splash Of  Lime 175g         1        4.4  
264833                  Doritos Mexicana    170g         2        8.8  
264834   Doritos Corn Chip Mexican Jalapeno 150g         2        7.8  
264835             Tostitos Splash Of  Lime 175g         2        8.8  

[264836 rows x 8 columns]

Dataset 2: Transaction Data (Individual Purchases)

| Column Name        | Description |
|--------------------|-------------|
| `DATE`             | Purchase date in Excel serial number format (e.g., `43390`). |
| `STORE_NBR`        | Store number where the transaction took place. |
| `LYLTY_CARD_NBR`   | Loyalty card number (same as above). |
| `TXN_ID`           | Unique transaction ID for identifying a specific purchase event. |
| `PROD_NBR`         | Internal product ID used for inventory and categorization. |
| `PROD_NAME`        | Full product name (e.g., "Smiths Crinkle Cut Chips Chicken 170g"). Includes brand, flavor, size. |
| `PROD_QTY`         | Number of units purchased of the product in a single transaction. |
| `TOT_SALES`        | Total cost for that product line in the transaction. |

---

Relationship Between the Datasets

Can join both datasets using the `LYLTY_CARD_NBR` column

#### Key Insights to Extract from the Dataset

1. **Product Popularity by Life Stage**
 
    Understand which products are most preferred by different life stages such as:
    - YOUNG FAMILIES
    - OLDER SINGLES/COUPLES
    - RETIREES

    Helps in targeted marketing — e.g., promoting family-size packs to young families or smaller snack packs to older singles.

---

2. **Product Popularity Among Premium Customers**

   Find out which products are mostly bought by **Premium** spenders vs **Budget** or **Mainstream** segments.
   Brands can:
   - Upsell premium products to high-value customers

---

3. **Sales Across Different Stores**

    Compare total sales across different `STORE_NBR`s.
 
    - Identify high-performing and underperforming stores
    - Detect location-based preferences

---

4. **Product vs Total Sales**
   
   Aggregate `TOT_SALES` for each `PROD_NAME` to determine revenue-driving products.

   - Identify bestsellers

---

5. **Life Stage vs Quantity Purchased**

   Helps in knowing which LIFESTYLE buy more.

   - Can make packages which attract a particular category to increase sales.

---



# STEP 2: CLEANING AND MERGING THE DATA

## Purchase Behaviour Data (Customer Segment)

In [ ]:
purchase_df

LYLTY_CARD_NBR               LIFESTAGE PREMIUM_CUSTOMER
0                1000   YOUNG SINGLES/COUPLES          Premium
1                1002   YOUNG SINGLES/COUPLES       Mainstream
2                1003          YOUNG FAMILIES           Budget
3                1004   OLDER SINGLES/COUPLES       Mainstream
4                1005  MIDAGE SINGLES/COUPLES       Mainstream
...               ...                     ...              ...
72632         2370651  MIDAGE SINGLES/COUPLES       Mainstream
72633         2370701          YOUNG FAMILIES       Mainstream
72634         2370751          YOUNG FAMILIES          Premium
72635         2370961          OLDER FAMILIES           Budget
72636         2373711   YOUNG SINGLES/COUPLES       Mainstream

[72637 rows x 3 columns]

In [ ]:
purchase_df['LIFESTAGE'] = purchase_df['LIFESTAGE'].str.title()

Check for missing values

In [ ]:
purchase_df.isnull().sum()

LYLTY_CARD_NBR      0
LIFESTAGE           0
PREMIUM_CUSTOMER    0
dtype: int64

Checking Datatype

In [ ]:
purchase_df.dtypes

LYLTY_CARD_NBR       int64
LIFESTAGE           object
PREMIUM_CUSTOMER    object
dtype: object

Checking for duplicates

In [ ]:
print(purchase_df["LYLTY_CARD_NBR"].value_counts())

LYLTY_CARD_NBR
1000       1
181211     1
181217     1
181216     1
181215     1
          ..
88126      1
88127      1
88128      1
88129      1
2373711    1
Name: count, Length: 72637, dtype: int64


In [ ]:
purchase_df = purchase_df.drop_duplicates(keep='last')

Checking the values in categorical columns

In [ ]:
# Filter columns with dtype 'object'
object_columns = purchase_df.select_dtypes(include='object')

# Print unique values for each object column
for col in object_columns.columns:
    print(f"\nColumn: {col}")
    print(purchase_df[col].unique())


Column: LIFESTAGE
['Young Singles/Couples' 'Young Families' 'Older Singles/Couples'
 'Midage Singles/Couples' 'New Families' 'Older Families' 'Retirees']

Column: PREMIUM_CUSTOMER
['Premium' 'Mainstream' 'Budget']


To get the better insight, I split the LIFESTYLE column into AGE_GROUP and FAMILY_STATUS

- AGE_GROUP will contain Young, older and midage

- FAMILY_STATUS will contain Singles/Couples, Families, Retirees

In [ ]:
import numpy as np

# Create empty columns
purchase_df['AGE_GROUP'] = np.nan
purchase_df['FAMILY_STATUS'] = np.nan

# Case 1: For 'RETIREES'
mask_retirees = purchase_df['LIFESTAGE'] == 'Retirees'
purchase_df.loc[mask_retirees, 'FAMILY_STATUS'] = 'Retirees'

# Case 2: For others (split into age group and family status)
mask_others = ~mask_retirees
split_df = purchase_df.loc[mask_others, 'LIFESTAGE'].str.split(' ', n=1, expand=True)

# Assign the split parts
purchase_df.loc[mask_others, 'AGE_GROUP'] = split_df[0]
purchase_df.loc[mask_others, 'FAMILY_STATUS'] = split_df[1]

In [ ]:
purchase_df

LYLTY_CARD_NBR               LIFESTAGE PREMIUM_CUSTOMER AGE_GROUP  \
0                1000   Young Singles/Couples          Premium     Young   
1                1002   Young Singles/Couples       Mainstream     Young   
2                1003          Young Families           Budget     Young   
3                1004   Older Singles/Couples       Mainstream     Older   
4                1005  Midage Singles/Couples       Mainstream    Midage   
...               ...                     ...              ...       ...   
72632         2370651  Midage Singles/Couples       Mainstream    Midage   
72633         2370701          Young Families       Mainstream     Young   
72634         2370751          Young Families          Premium     Young   
72635         2370961          Older Families           Budget     Older   
72636         2373711   Young Singles/Couples       Mainstream     Young   

         FAMILY_STATUS  
0      Singles/Couples  
1      Singles/Couples  
2             Families  
3      Singles/Couples  
4      Singles/Couples  
...                ...  
72632  Singles/Couples  
72633         Families  
72634         Families  
72635         Families  
72636  Singles/Couples  

[72637 rows x 5 columns]

In [ ]:
purchase_df.isnull().sum()

LYLTY_CARD_NBR          0
LIFESTAGE               0
PREMIUM_CUSTOMER        0
AGE_GROUP           14805
FAMILY_STATUS           0
dtype: int64

In [ ]:
purchase_df["AGE_GROUP"] = purchase_df["AGE_GROUP"].fillna("RETIRED")

In [ ]:
purchase_df.isnull().sum()

LYLTY_CARD_NBR      0
LIFESTAGE           0
PREMIUM_CUSTOMER    0
AGE_GROUP           0
FAMILY_STATUS       0
dtype: int64

## Transaction Data

In [ ]:
trans_df

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0       43390          1            1000       1         5   
1       43599          1            1307     348        66   
2       43605          1            1343     383        61   
3       43329          2            2373     974        69   
4       43330          2            2426    1038       108   
...       ...        ...             ...     ...       ...   
264831  43533        272          272319  270088        89   
264832  43325        272          272358  270154        74   
264833  43410        272          272379  270187        51   
264834  43461        272          272379  270188        42   
264835  43365        272          272380  270189        74   

                                       PROD_NAME  PROD_QTY  TOT_SALES  
0         Natural Chip        Compny SeaSalt175g         2        6.0  
1                       CCs Nacho Cheese    175g         3        6.3  
2         Smiths Crinkle Cut  Chips Chicken 170g         2        2.9  
3         Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0  
4       Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8  
...                                          ...       ...        ...  
264831   Kettle Sweet Chilli And Sour Cream 175g         2       10.8  
264832             Tostitos Splash Of  Lime 175g         1        4.4  
264833                  Doritos Mexicana    170g         2        8.8  
264834   Doritos Corn Chip Mexican Jalapeno 150g         2        7.8  
264835             Tostitos Splash Of  Lime 175g         2        8.8  

[264836 rows x 8 columns]

Checking for missing values

In [ ]:
trans_df.isnull().sum()

DATE              0
STORE_NBR         0
LYLTY_CARD_NBR    0
TXN_ID            0
PROD_NBR          0
PROD_NAME         0
PROD_QTY          0
TOT_SALES         0
dtype: int64

Checking for duplicates

In [ ]:
trans_df.shape

(264836, 8)

In [ ]:
trans_df = trans_df.drop_duplicates(keep='last')

In [ ]:
trans_df.shape

(264835, 8)

Checking datatypes

In [ ]:
trans_df.dtypes

DATE                int64
STORE_NBR           int64
LYLTY_CARD_NBR      int64
TXN_ID              int64
PROD_NBR            int64
PROD_NAME          object
PROD_QTY            int64
TOT_SALES         float64
dtype: object

Date should be in DateTime not in int

In [ ]:
trans_df['DATE'] = pd.to_datetime(
    trans_df['DATE'],
    origin='1899-12-30',
    unit='D'
).dt.date

trans_df['DATE'] = pd.to_datetime(trans_df['DATE'])


C:\Users\HP\AppData\Local\Temp\ipykernel_23188\17482334.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  trans_df['DATE'] = pd.to_datetime(
C:\Users\HP\AppData\Local\Temp\ipykernel_23188\17482334.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  trans_df['DATE'] = pd.to_datetime(trans_df['DATE'])


In [ ]:
trans_df

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0      2018-10-17          1            1000       1         5   
1      2019-05-14          1            1307     348        66   
2      2019-05-20          1            1343     383        61   
3      2018-08-17          2            2373     974        69   
4      2018-08-18          2            2426    1038       108   
...           ...        ...             ...     ...       ...   
264831 2019-03-09        272          272319  270088        89   
264832 2018-08-13        272          272358  270154        74   
264833 2018-11-06        272          272379  270187        51   
264834 2018-12-27        272          272379  270188        42   
264835 2018-09-22        272          272380  270189        74   

                                       PROD_NAME  PROD_QTY  TOT_SALES  
0         Natural Chip        Compny SeaSalt175g         2        6.0  
1                       CCs Nacho Cheese    175g         3        6.3  
2         Smiths Crinkle Cut  Chips Chicken 170g         2        2.9  
3         Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0  
4       Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8  
...                                          ...       ...        ...  
264831   Kettle Sweet Chilli And Sour Cream 175g         2       10.8  
264832             Tostitos Splash Of  Lime 175g         1        4.4  
264833                  Doritos Mexicana    170g         2        8.8  
264834   Doritos Corn Chip Mexican Jalapeno 150g         2        7.8  
264835             Tostitos Splash Of  Lime 175g         2        8.8  

[264835 rows x 8 columns]

In [ ]:
print(trans_df["PROD_NAME"].unique())
print(trans_df["PROD_NAME"].nunique())

['Natural Chip        Compny SeaSalt175g' 'CCs Nacho Cheese    175g'
 'Smiths Crinkle Cut  Chips Chicken 170g'
 'Smiths Chip Thinly  S/Cream&Onion 175g'
 'Kettle Tortilla ChpsHny&Jlpno Chili 150g'
 'Old El Paso Salsa   Dip Tomato Mild 300g'
 'Smiths Crinkle Chips Salt & Vinegar 330g'
 'Grain Waves         Sweet Chilli 210g'
 'Doritos Corn Chip Mexican Jalapeno 150g'
 'Grain Waves Sour    Cream&Chives 210G'
 'Kettle Sensations   Siracha Lime 150g' 'Twisties Cheese     270g'
 'WW Crinkle Cut      Chicken 175g' 'Thins Chips Light&  Tangy 175g'
 'CCs Original 175g' 'Burger Rings 220g'
 'NCC Sour Cream &    Garden Chives 175g'
 'Doritos Corn Chip Southern Chicken 150g' 'Cheezels Cheese Box 125g'
 'Smiths Crinkle      Original 330g'
 'Infzns Crn Crnchers Tangy Gcamole 110g'
 'Kettle Sea Salt     And Vinegar 175g'
 'Smiths Chip Thinly  Cut Original 175g' 'Kettle Original 175g'
 'Red Rock Deli Thai  Chilli&Lime 150g' 'Pringles Sthrn FriedChicken 134g'
 'Pringles Sweet&Spcy BBQ 134g' 'Red Rock 

Product name have typos and is inconsistent

It is mix of **brandname + Flavor + weight**

For better insights, can be made three different columns

Correcting typos

In [ ]:
import re

trans_df['PROD_NAME'] = trans_df['PROD_NAME'].str.replace(r'\s+', ' ', regex=True).str.strip()

# Step 2: Replace known typos
replacements = {
    'Compny': 'Company',
    'Companympany': 'Company',
    'Crn Crnchers': 'Corn Crunchers',
    'Crnkle': 'Crinkle',
    'Crnkle Cut': 'Crinkle Cut',
    'Crnkle Chip': 'Crinkle Cut',
    'Crinkle Chip': 'Crinkle Cut',
    'Orgnl': 'Original',
    'Originl': 'Original',
    'OriginalBig Bag': 'Original Big Bag',
    'Chikn': 'Chicken',
    'Chckn': 'Chicken',
    'ChpsHny&Jlpno': 'Chips Honey & Jalapeno',
    'ChpsBtroot&Ricotta': 'Chips Beetroot & Ricotta',
    'ChpsFeta&Garlic': 'Chips Feta & Garlic',
    'Swt Chli&S/Cream175G': 'Sweet Chilli & Sour Cream 175g',
    'Swt/Chlli &Sr/Cream': 'Sweet Chilli & Sour Cream',
    'Seasonedchicken': 'Seasoned Chicken',
    'Tmato Hrb&Spce': 'Tomato Herb & Spice',
    'Tangy Gcamole': 'Tangy Guacamole',
    'Tangy Guacamole': 'Tangy Guacamole',
    'Chutny': 'Chutney',
    'Cheddr&Mstrd': 'Cheddar & Mustard',
    'Ched & Chives': 'Cheddar & Chives',
    'BBQ Rib': 'BBQ Ribs',
    'BBQ Ribss': 'BBQ Ribs',
    'OnionStacked': 'Onion Stacked',
    'Pc Sea Salt': 'Sea Salt',
    'Slt Vingar': 'Salt Vinegar',
    'Salt/Vinegr': 'Salt & Vinegar',
    'Salt & Vinegr': 'Salt & Vinegar',
    'SR Slow Rst': 'Slow Roasted',
    'Slow Rst Pork Belly': 'Slow Roasted Pork Belly',
    'Frch/Onin': 'French Onion',
    'French OnionDip': 'French Onion Dip',
    'Whlegrn': 'Wholegrain',
    'Wholegrain': 'Wholegrain', 
    'Chnky Tom Ht300g': 'Chunky Tomato Hot 300g',
    'Chunky Tomato Hot': 'Chunky Tomato Hot',
    'Med': 'Medium',
    'Mediumiumium': 'Medium',
    'Mediumium': 'Medium',
    'Honey Soy Chicken175g': 'Honey Soy Chicken 175g',
    'Hony': 'Honey',
    'Hony Soy': 'Honey Soy',
    'Thins Chips': 'Thins',
    'Thins Potato Chips': 'Thins',
    'Smith Crinkle Cut': 'Smiths Crinkle Cut',
    'Smiths Chip Thinly': 'Smiths Thinly Cut',
    'Smiths Chip Thinly CutSalt/Vinegr': 'Smiths Thinly Cut Salt & Vinegar',
    'GrnWves': 'Grain Waves',
    'GrainWves': 'Grain Waves',
    'RRD': 'Red Rock Deli',
    'Snbts': 'Sunbites',
    'Sunbites': 'Sunbites',
    'Chps': 'Chips',
    'Chp': 'Chip',
    'Crips': 'Crisps',
    'Swt Pot': 'Sweet Potato',
    'Nacho Cheese170g': 'Nacho Cheese 170g',
    'Supreme380g': 'Supreme 380g',
    'Cheddar&Mstrd': 'Cheddar & Mustard',
    'SeaSalt': 'Sea Salt',
    'Light& Tangy': 'Light & Tangy',
    'Light &Tangy': 'Light & Tangy',
    'Salt/Vinegr': 'Salt & Vinegar',
    'Original saltd': 'Original Salted',
    'OnionDip': 'Onion Dip',
    'Seasonedchicken': 'Seasoned Chicken',
    'Mzzrlla': 'Mozzarella',
    'Btroot': 'Beetroot',
    'Chs&Onion': 'Cheese & Onion',
    'Tom Ht': 'Tomato Hot',
    'Tomato Med': 'Tomato Medium',
    'OriginalBig Bag': 'Original Big Bag',
    'Mystery Flavour': 'Mystery Flavour',
    'FriedChicken': 'Fried Chicken',
    'Chimuchurri': 'Chimichurri',
    'Mediumiumium': 'Medium',
    'Mediumium': 'Medium',
    'SourCream&Herbs': 'Sour Cream & Herbs',
    'Sour Crm &Chives': 'Sour Cream & Chives',
    'Salt Crips': 'Salt Crisps',
    'Crinkle Cut Mac N Cheese': 'Crinkle Cut Mac N Cheese',
    'Infzns':'Infuzions', 
    'Natural ChipCo' : 'Natural Chip Company',
    'Dorito': 'Doritos',
    'Doritoss': 'Doritos', 
    'Natural Chip Co' : 'Natural Chip Company',
    'Natural Chip Companympany' : 'Natural Chip Company'
}


for wrong, correct in replacements.items():
    trans_df['PROD_NAME'] = trans_df['PROD_NAME'].str.replace(wrong, correct, regex=False)


C:\Users\HP\AppData\Local\Temp\ipykernel_23188\3160802690.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  trans_df['PROD_NAME'] = trans_df['PROD_NAME'].str.replace(r'\s+', ' ', regex=True).str.strip()
C:\Users\HP\AppData\Local\Temp\ipykernel_23188\3160802690.py:101: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  trans_df['PROD_NAME'] = trans_df['PROD_NAME'].str.replace(wrong, correct, regex=False)


In [ ]:
trans_df

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0      2018-10-17          1            1000       1         5   
1      2019-05-14          1            1307     348        66   
2      2019-05-20          1            1343     383        61   
3      2018-08-17          2            2373     974        69   
4      2018-08-18          2            2426    1038       108   
...           ...        ...             ...     ...       ...   
264831 2019-03-09        272          272319  270088        89   
264832 2018-08-13        272          272358  270154        74   
264833 2018-11-06        272          272379  270187        51   
264834 2018-12-27        272          272379  270188        42   
264835 2018-09-22        272          272380  270189        74   

                                                PROD_NAME  PROD_QTY  TOT_SALES  
0                       Natural Chip Company Sea Salt175g         2        6.0  
1                                   CCs Nacho Cheese 175g         3        6.3  
2                   Smiths Crinkle Cut Chips Chicken 170g         2        2.9  
3                    Smiths Thinly Cut S/Cream&Onion 175g         5       15.0  
4       Kettle Tortilla Chips Honey & Jalapeno Chili 150g         3       13.8  
...                                                   ...       ...        ...  
264831            Kettle Sweet Chilli And Sour Cream 175g         2       10.8  
264832                       Tostitos Splash Of Lime 175g         1        4.4  
264833                              Doritos Mexicana 170g         2        8.8  
264834            Doritos Corn Chip Mexican Jalapeno 150g         2        7.8  
264835                       Tostitos Splash Of Lime 175g         2        8.8  

[264835 rows x 8 columns]

In [ ]:
print(trans_df["PROD_NAME"].unique())

['Natural Chip Company Sea Salt175g' 'CCs Nacho Cheese 175g'
 'Smiths Crinkle Cut Chips Chicken 170g'
 'Smiths Thinly Cut S/Cream&Onion 175g'
 'Kettle Tortilla Chips Honey & Jalapeno Chili 150g'
 'Old El Paso Salsa Dip Tomato Mild 300g'
 'Smiths Crinkle Cuts Salt & Vinegar 330g' 'Grain Waves Sweet Chilli 210g'
 'Doritos Corn Chip Mexican Jalapeno 150g'
 'Grain Waves Sour Cream&Chives 210G'
 'Kettle Sensations Siracha Lime 150g' 'Twisties Cheese 270g'
 'WW Crinkle Cut Chicken 175g' 'Thins Light & Tangy 175g'
 'CCs Original 175g' 'Burger Rings 220g'
 'NCC Sour Cream & Garden Chives 175g'
 'Doritos Corn Chip Southern Chicken 150g' 'Cheezels Cheese Box 125g'
 'Smiths Crinkle Original 330g'
 'Infuzions Corn Crunchers Tangy Guacamole 110g'
 'Kettle Sea Salt And Vinegar 175g' 'Smiths Thinly Cut Cut Original 175g'
 'Kettle Original 175g' 'Red Rock Deli Thai Chilli&Lime 150g'
 'Pringles Sthrn Fried Chicken 134g' 'Pringles Sweet&Spcy BBQ 134g'
 'Red Rock Deli SR Salsa & Mozzarella 150g' 'Thins O

In [ ]:
trans_df['PROD_NAME'] = trans_df['PROD_NAME'].str.replace(r'\s+', ' ', regex=True).str.strip()

C:\Users\HP\AppData\Local\Temp\ipykernel_23188\624741088.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  trans_df['PROD_NAME'] = trans_df['PROD_NAME'].str.replace(r'\s+', ' ', regex=True).str.strip()


Extract Weight

In [ ]:
# Extract weight with 'g' or 'G'
trans_df['WEIGHT'] = trans_df['PROD_NAME'].str.extract(r'(\d{2,4}[gG])', expand=False)

# Remove the weight from product name and clean up the rest
trans_df['NO_WEIGHT'] = trans_df['PROD_NAME'].str.replace(r'\d{2,4}[gG]', '', regex=True).str.strip()


C:\Users\HP\AppData\Local\Temp\ipykernel_23188\943940521.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  trans_df['WEIGHT'] = trans_df['PROD_NAME'].str.extract(r'(\d{2,4}[gG])', expand=False)
C:\Users\HP\AppData\Local\Temp\ipykernel_23188\943940521.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  trans_df['NO_WEIGHT'] = trans_df['PROD_NAME'].str.replace(r'\d{2,4}[gG]', '', regex=True).str.strip()


Extract Brand

In [ ]:
brand_list = [
    'Smiths', 'Natural Chip Company', 'Kettle', 'Red Rock Deli', 'Doritos',
    'Twisties', 'Old El Paso', 'Grain Waves', 'CCs', 'Thins', 'Pringles',
    'Cobs', 'Tostitos', 'Cheezels', 'Infuzions', 'Tyrrells', 'Snbts',
    'Sunbites', 'Cheetos', 'Woolworths', 'WW', 'NCC', 'RRD', 'French Fries','Burger Rings'
    
]    

def extract_brand(product_name):
    for brand in sorted(brand_list, key=len, reverse=True):
        if product_name.strip().startswith(brand):
            return brand
    return 'Unknown'


trans_df['BRAND'] = trans_df['NO_WEIGHT'].apply(extract_brand)


C:\Users\HP\AppData\Local\Temp\ipykernel_23188\53823019.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  trans_df['BRAND'] = trans_df['NO_WEIGHT'].apply(extract_brand)


In [ ]:
len(trans_df[trans_df['BRAND'] == 'Unknown'])

0

Extract Flavor

In [ ]:
trans_df['FLAVOR'] = trans_df.apply(lambda row: row['NO_WEIGHT'].replace(row['BRAND'], '').strip(), axis=1)

C:\Users\HP\AppData\Local\Temp\ipykernel_23188\422548400.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  trans_df['FLAVOR'] = trans_df.apply(lambda row: row['NO_WEIGHT'].replace(row['BRAND'], '').strip(), axis=1)


In [ ]:
trans_df.loc[trans_df['BRAND'] == 'Burger Rings', 'FLAVOR'] = 'Burger Rings'

In [ ]:
trans_df.drop(columns=['NO_WEIGHT'], inplace=True)

C:\Users\HP\AppData\Local\Temp\ipykernel_23188\3770478242.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  trans_df.drop(columns=['NO_WEIGHT'], inplace=True)


In [ ]:
trans_df

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0      2018-10-17          1            1000       1         5   
1      2019-05-14          1            1307     348        66   
2      2019-05-20          1            1343     383        61   
3      2018-08-17          2            2373     974        69   
4      2018-08-18          2            2426    1038       108   
...           ...        ...             ...     ...       ...   
264831 2019-03-09        272          272319  270088        89   
264832 2018-08-13        272          272358  270154        74   
264833 2018-11-06        272          272379  270187        51   
264834 2018-12-27        272          272379  270188        42   
264835 2018-09-22        272          272380  270189        74   

                                                PROD_NAME  PROD_QTY  \
0                       Natural Chip Company Sea Salt175g         2   
1                                   CCs Nacho Cheese 175g         3   
2                   Smiths Crinkle Cut Chips Chicken 170g         2   
3                    Smiths Thinly Cut S/Cream&Onion 175g         5   
4       Kettle Tortilla Chips Honey & Jalapeno Chili 150g         3   
...                                                   ...       ...   
264831            Kettle Sweet Chilli And Sour Cream 175g         2   
264832                       Tostitos Splash Of Lime 175g         1   
264833                              Doritos Mexicana 170g         2   
264834            Doritos Corn Chip Mexican Jalapeno 150g         2   
264835                       Tostitos Splash Of Lime 175g         2   

        TOT_SALES WEIGHT                 BRAND  \
0             6.0   175g  Natural Chip Company   
1             6.3   175g                   CCs   
2             2.9   170g                Smiths   
3            15.0   175g                Smiths   
4            13.8   150g                Kettle   
...           ...    ...                   ...   
264831       10.8   175g                Kettle   
264832        4.4   175g              Tostitos   
264833        8.8   170g               Doritos   
264834        7.8   150g               Doritos   
264835        8.8   175g              Tostitos   

                                       FLAVOR  
0                                    Sea Salt  
1                                Nacho Cheese  
2                   Crinkle Cut Chips Chicken  
3                    Thinly Cut S/Cream&Onion  
4       Tortilla Chips Honey & Jalapeno Chili  
...                                       ...  
264831            Sweet Chilli And Sour Cream  
264832                         Splash Of Lime  
264833                               Mexicana  
264834             Corn Chip Mexican Jalapeno  
264835                         Splash Of Lime  

[264835 rows x 11 columns]

In [ ]:
trans_df.isnull().sum()

DATE              0
STORE_NBR         0
LYLTY_CARD_NBR    0
TXN_ID            0
PROD_NBR          0
PROD_NAME         0
PROD_QTY          0
TOT_SALES         0
WEIGHT            0
BRAND             0
FLAVOR            0
dtype: int64

In [ ]:
trans_df.drop(columns=['PROD_NAME'], inplace=True)

C:\Users\HP\AppData\Local\Temp\ipykernel_23188\3166844902.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  trans_df.drop(columns=['PROD_NAME'], inplace=True)


In [ ]:
trans_df

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  PROD_QTY  \
0      2018-10-17          1            1000       1         5         2   
1      2019-05-14          1            1307     348        66         3   
2      2019-05-20          1            1343     383        61         2   
3      2018-08-17          2            2373     974        69         5   
4      2018-08-18          2            2426    1038       108         3   
...           ...        ...             ...     ...       ...       ...   
264831 2019-03-09        272          272319  270088        89         2   
264832 2018-08-13        272          272358  270154        74         1   
264833 2018-11-06        272          272379  270187        51         2   
264834 2018-12-27        272          272379  270188        42         2   
264835 2018-09-22        272          272380  270189        74         2   

        TOT_SALES WEIGHT                 BRAND  \
0             6.0   175g  Natural Chip Company   
1             6.3   175g                   CCs   
2             2.9   170g                Smiths   
3            15.0   175g                Smiths   
4            13.8   150g                Kettle   
...           ...    ...                   ...   
264831       10.8   175g                Kettle   
264832        4.4   175g              Tostitos   
264833        8.8   170g               Doritos   
264834        7.8   150g               Doritos   
264835        8.8   175g              Tostitos   

                                       FLAVOR  
0                                    Sea Salt  
1                                Nacho Cheese  
2                   Crinkle Cut Chips Chicken  
3                    Thinly Cut S/Cream&Onion  
4       Tortilla Chips Honey & Jalapeno Chili  
...                                       ...  
264831            Sweet Chilli And Sour Cream  
264832                         Splash Of Lime  
264833                               Mexicana  
264834             Corn Chip Mexican Jalapeno  
264835                         Splash Of Lime  

[264835 rows x 10 columns]

In [ ]:
trans_df.dtypes

DATE              datetime64[ns]
STORE_NBR                  int64
LYLTY_CARD_NBR             int64
TXN_ID                     int64
PROD_NBR                   int64
PROD_QTY                   int64
TOT_SALES                float64
WEIGHT                    object
BRAND                     object
FLAVOR                    object
dtype: object

In [ ]:
print('Weights:\n')
print(trans_df['WEIGHT'].str.lower().unique())
print("\nNo. of unique values:", trans_df['WEIGHT'].str.lower().nunique(),'\n')
print('\nBrands:\n')
print(trans_df['BRAND'].str.lower().unique())
print("\nNo. of unique values:", trans_df['BRAND'].str.lower().nunique(),'\n')
print('\nFlavors:\n')
print(trans_df['FLAVOR'].str.lower().unique())
print("\nNo. of unique values:", trans_df['FLAVOR'].str.lower().nunique(),'\n')


Weights:

['175g' '170g' '150g' '300g' '330g' '210g' '270g' '220g' '125g' '110g'
 '134g' '380g' '180g' '165g' '135g' '250g' '200g' '160g' '190g' '90g'
 '70g']

No. of unique values: 21 


Brands:

['natural chip company' 'ccs' 'smiths' 'kettle' 'old el paso'
 'grain waves' 'doritos' 'twisties' 'ww' 'thins' 'burger rings' 'ncc'
 'cheezels' 'infuzions' 'red rock deli' 'pringles' 'tyrrells' 'cobs'
 'woolworths' 'french fries' 'tostitos' 'cheetos' 'sunbites']

No. of unique values: 23 


Flavors:

['sea salt' 'nacho cheese' 'crinkle cut chips chicken'
 'thinly cut s/cream&onion' 'tortilla chips honey & jalapeno chili'
 'salsa dip tomato mild' 'crinkle cuts salt & vinegar' 'sweet chilli'
 'corn chip mexican jalapeno' 'sour cream&chives'
 'sensations siracha lime' 'cheese' 'crinkle cut chicken' 'light & tangy'
 'original' 'burger rings' 'sour cream & garden chives'
 'corn chip southern chicken' 'cheese box' 'crinkle original'
 'corn crunchers tangy guacamole' 'sea salt and vinegar'
 'thinly 

## Merging Datasets

Now Both transaction data and customer data are ready to merge

In [ ]:
trans_df.columns

Index(['DATE', 'STORE_NBR', 'LYLTY_CARD_NBR', 'TXN_ID', 'PROD_NBR', 'PROD_QTY',
       'TOT_SALES', 'WEIGHT', 'BRAND', 'FLAVOR'],
      dtype='object')

In [ ]:
merged_df = pd.merge(trans_df, purchase_df, on='LYLTY_CARD_NBR', how='inner')

In [ ]:
merged_df

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  PROD_QTY  \
0      2018-10-17          1            1000       1         5         2   
1      2019-05-14          1            1307     348        66         3   
2      2018-11-10          1            1307     346        96         2   
3      2019-03-09          1            1307     347        54         1   
4      2019-05-20          1            1343     383        61         2   
...           ...        ...             ...     ...       ...       ...   
264830 2019-03-09        272          272319  270088        89         2   
264831 2018-08-13        272          272358  270154        74         1   
264832 2018-11-06        272          272379  270187        51         2   
264833 2018-12-27        272          272379  270188        42         2   
264834 2018-09-22        272          272380  270189        74         2   

        TOT_SALES WEIGHT                 BRAND                       FLAVOR  \
0             6.0   175g  Natural Chip Company                     Sea Salt   
1             6.3   175g                   CCs                 Nacho Cheese   
2             3.8   160g                    WW       Original Stacked Chips   
3             2.1   175g                   CCs                     Original   
4             2.9   170g                Smiths    Crinkle Cut Chips Chicken   
...           ...    ...                   ...                          ...   
264830       10.8   175g                Kettle  Sweet Chilli And Sour Cream   
264831        4.4   175g              Tostitos               Splash Of Lime   
264832        8.8   170g               Doritos                     Mexicana   
264833        7.8   150g               Doritos   Corn Chip Mexican Jalapeno   
264834        8.8   175g              Tostitos               Splash Of Lime   

                     LIFESTAGE PREMIUM_CUSTOMER AGE_GROUP    FAMILY_STATUS  
0        Young Singles/Couples          Premium     Young  Singles/Couples  
1       Midage Singles/Couples           Budget    Midage  Singles/Couples  
2       Midage Singles/Couples           Budget    Midage  Singles/Couples  
3       Midage Singles/Couples           Budget    Midage  Singles/Couples  
4       Midage Singles/Couples           Budget    Midage  Singles/Couples  
...                        ...              ...       ...              ...  
264830   Young Singles/Couples          Premium     Young  Singles/Couples  
264831   Young Singles/Couples          Premium     Young  Singles/Couples  
264832   Young Singles/Couples          Premium     Young  Singles/Couples  
264833   Young Singles/Couples          Premium     Young  Singles/Couples  
264834   Young Singles/Couples          Premium     Young  Singles/Couples  

[264835 rows x 14 columns]

In [ ]:
merged_df.isnull().sum()

DATE                0
STORE_NBR           0
LYLTY_CARD_NBR      0
TXN_ID              0
PROD_NBR            0
PROD_QTY            0
TOT_SALES           0
WEIGHT              0
BRAND               0
FLAVOR              0
LIFESTAGE           0
PREMIUM_CUSTOMER    0
AGE_GROUP           0
FAMILY_STATUS       0
dtype: int64